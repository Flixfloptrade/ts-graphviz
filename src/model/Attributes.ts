import { DotBase } from '../common';
import { indent } from '../utils/dot-rendering';
import { Literal } from './Literal';

/**
 * @category Attributes
 */
export class Attributes extends DotBase {
  protected attrs: Map<string, Literal> = new Map();
  get size(): number {
    return this.attrs.size;
  }
  public get(key: string): Literal | undefined {
    return this.attrs.get(key);
  }
  public set(key: string, value: any): void {
    if (value instanceof Literal) {
      this.attrs.set(key, value);
    } else {
      this.attrs.set(key, new Literal(value));
    }
  }
  public toDot(): string {
    if (this.size === 0) {
      return '';
    }
    return [
      '[',
      ...Array.from(this.attrs.entries()).map(([key, value]) => indent(`${key} = ${value.toDot()},`)),
      ']',
    ].join('\n');
  }
}
