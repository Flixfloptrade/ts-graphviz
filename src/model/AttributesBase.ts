import { DotBase } from '../abstract';
import { AttributesObject, AttributesValue, IAttributesBase } from '../types';

/**
 * @hidden
 */
export abstract class AttributesBase<T extends string> extends DotBase implements IAttributesBase<T> {
  /** @hidden */
  protected attrs: Map<T, AttributesValue> = new Map();

  public entries(): IterableIterator<[T, AttributesValue]> {
    return this.attrs.entries();
  }

  /** The size of the attribute. */
  get size(): number {
    return this.attrs.size;
  }
  /** The size of the attribute. */
  public get(key: T): AttributesValue | undefined {
    return this.attrs.get(key);
  }
  /** Set a value to the attribute. */
  public set(key: T, value: AttributesValue): void {
    this.attrs.set(key, value);
  }

  public delete(key: T): void {
    this.attrs.delete(key);
  }

  public apply(attributes: AttributesObject<T>): void {
    for (const [key, value] of Object.entries(attributes)) {
      this.set(key as T, value as string | boolean | number);
    }
  }

  public clear(): void {
    this.attrs.clear();
  }
}
