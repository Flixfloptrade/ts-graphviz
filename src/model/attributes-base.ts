import { DotObject } from './abstract';
import { AttributesObject, AttributesValue, IAttributesBase, IAttributes } from '../types';

/**
 * @hidden
 */
export abstract class AttributesBase<T extends string> extends DotObject implements IAttributesBase<T> {
  /** @hidden */
  protected attrs: Map<T, AttributesValue> = new Map();

  public entries(): IterableIterator<[T, AttributesValue]> {
    return this.attrs.entries();
  }

  /** The size of the attribute. */
  get size(): number {
    return this.attrs.size;
  }
  /** The size of the attribute. */
  public get(key: T): AttributesValue | undefined {
    return this.attrs.get(key);
  }
  /** Set a value to the attribute. */
  public set(key: T, value: AttributesValue): void {
    this.attrs.set(key, value);
  }

  public delete(key: T): void {
    this.attrs.delete(key);
  }

  public apply(attributes: AttributesObject<T>): void {
    for (const [key, value] of Object.entries(attributes)) {
      this.set(key as T, value as string | boolean | number);
    }
  }

  public clear(): void {
    this.attrs.clear();
  }
}

/**
 * A set of attribute values for any object.
 *
 * @category Attributes
 */
export class Attributes<T extends string = string> extends AttributesBase<T> implements IAttributes<T> {
  /** Comments to include when outputting with toDot. */
  public comment?: string;
}
