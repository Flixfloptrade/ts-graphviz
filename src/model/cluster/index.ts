export * from './Cluster';
export * from './RootCluster';
export * from './Graph';
export * from './Digraph';
