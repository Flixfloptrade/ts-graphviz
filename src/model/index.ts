export * from './Attributes';
export * from './Cluster';
export * from './RootCluster';
export * from './Graph';
export * from './Digraph';
export * from './Edge';
export * from './Node';
export * from '../render/Context';
export * from '../render/to-dot';
