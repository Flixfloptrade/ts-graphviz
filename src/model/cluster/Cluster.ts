import { DotBase } from '../../common';
import { SubgraphAttributes } from '../attributes';
import { Attributes } from '../attributes/Attributes';
import { EdgeAttributes } from '../attributes/EdgeAttributes';
import { NodeAttributes } from '../attributes/NodeAttributes';
import { Context } from '../Context';
import { DigraphEdge, Edge, GraphEdge } from '../Edge';
import { isNodeLike, isNodeLikeObject, Node, NodeLike, NodeLikeObject } from '../Node';

// tslint:disable: max-classes-per-file

export type RootClusterType = 'digraph' | 'graph';

export type ClusterType = RootClusterType | 'subgraph';
/**
 * @hidden
 */
export interface IClusterCommonAttributes<ATTR extends Attributes> {
  graph: ATTR;
  edge: EdgeAttributes;
  node: NodeAttributes;
}

/**
 * @hidden
 */
export abstract class Cluster<ATTR extends Attributes> extends DotBase {
  public abstract readonly context: Context;
  public abstract readonly type: ClusterType;
  public readonly attributes: Readonly<IClusterCommonAttributes<ATTR>>;

  private nodes: Map<string, Node> = new Map();

  private edges: Set<Edge> = new Set();

  private subgraphs: Map<string, Subgraph> = new Map();

  constructor(public readonly id: string, attributes: ATTR) {
    super();
    this.attributes = {
      graph: attributes,
      edge: new EdgeAttributes(),
      node: new NodeAttributes(),
    };
  }

  public add(object: Node | Edge | Subgraph): void {
    if (object instanceof Node) {
      this.addNode(object);
    } else if (object instanceof Edge) {
      this.addEdge(object);
    } else if (object instanceof Subgraph) {
      this.addSubgraph(object);
    }
  }

  public addNode(node: Node): void {
    this.nodes.set(node.id, node);
  }

  public addEdge(edge: Edge): void {
    this.edges.add(edge);
  }

  public addSubgraph(subgraph: Subgraph): void {
    this.subgraphs.set(subgraph.id, subgraph);
  }

  public existNode(nodeId: string): boolean {
    return this.nodes.has(nodeId);
  }

  public existEdge(edge: Edge): boolean {
    return this.edges.has(edge);
  }

  public existSubgraph(subgraphId: string): boolean {
    return this.subgraphs.has(subgraphId);
  }

  public createSubgraph(id: string): Subgraph {
    const graph = new Subgraph(this.context, id);
    this.subgraphs.set(id, graph);
    return graph;
  }

  public createNode(id: string): Node {
    const node = new Node(id);
    this.nodes.set(id, node);
    return node;
  }

  public getSubgraph(id: string): Subgraph | undefined {
    return this.subgraphs.get(id);
  }

  public getNode(id: string): Node | undefined {
    return this.nodes.get(id);
  }

  public createEdge(node1: NodeLike, node2: NodeLike, ...nodes: NodeLike[]): Edge;
  public createEdge(...nodes: NodeLike[]): Edge;
  public createEdge(node1: NodeLike, node2: NodeLike, ...nodes: NodeLike[]): Edge {
    if ((isNodeLike(node1) && isNodeLike(node2)) === false) {
      // TODO
      throw new Error();
    }

    const edge = new (this.context.graphType === 'graph' ? GraphEdge : DigraphEdge)(
      this.toNodeLikeObject(node1),
      this.toNodeLikeObject(node2),
      ...nodes.map(n => this.toNodeLikeObject(n)),
    );

    this.edges.add(edge);
    return edge;
  }

  public subgraph(id: string, callback: (subgraph: Subgraph) => void): Subgraph {
    const subgraph = this.getSubgraph(id) ?? this.createSubgraph(id);
    if (callback) {
      callback(subgraph);
    }
    return subgraph;
  }

  public node(id: string, callback?: (edge: Node) => void): Node {
    const node = this.getNode(id) ?? this.createNode(id);
    if (callback) {
      callback(node);
    }
    return node;
  }

  public edge(nodes: NodeLike[], callback?: (edge: Edge) => void): Edge {
    const edge = this.createEdge(...nodes);
    if (callback) {
      callback(edge);
    }
    return edge;
  }

  public toDot(): string {
    const type = this.type;
    const id = this.id;
    // attributes
    const commonAttributes = Object.entries(this.attributes)
      .filter(([_, attributes]) => attributes.size > 0)
      .map(([key, attributes]) => `${key} ${attributes.toDot()};`);

    // objects
    const nodes = Array.from(this.nodes.values()).map(o => o.toDot());
    const subgraphs = Array.from(this.subgraphs.values()).map(o => o.toDot());
    const edges = Array.from(this.edges.values()).map(o => o.toDot());
    const clusterContents = Cluster.joinLines(...commonAttributes, ...nodes, ...subgraphs, ...edges);
    const src = Cluster.joinLines(
      `${type} ${id} {`,
      clusterContents ? Cluster.indent(clusterContents) : undefined,
      '}',
    );
    return src;
  }

  private toNodeLikeObject(node: NodeLike): NodeLikeObject {
    if (isNodeLikeObject(node)) {
      return node;
    }
    // FIXME
    const [id, port] = node.split(':');
    const n = this.node(id);
    if (port) {
      return n.port(port);
    }
    return n;
  }
}

export abstract class RootCluster<ATTR extends Attributes> extends Cluster<ATTR> {
  public abstract readonly type: RootClusterType;
}

/**
 * @category Primary
 */
export class Subgraph extends Cluster<SubgraphAttributes> {
  public type: ClusterType = 'subgraph';

  constructor(public readonly context: Context, id: string, attributes: SubgraphAttributes = new SubgraphAttributes()) {
    super(id, attributes);
  }

  public isSubgraphCluster(): boolean {
    return this.id.startsWith('cluster_');
  }

  public toDot(): string {
    return `${super.toDot()};`;
  }
}
