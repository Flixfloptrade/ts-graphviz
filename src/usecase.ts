import { Context, Digraph, Graph, RootCluster } from './model';

/**
 * Type indicating that it is a constructor of T.
 * @hidden
 */
export type Type<T> = new (...args: any[]) => T;

/** @hidden */
const builder = <G extends RootCluster>(cls: Type<G>, strictMode: boolean = false) => (
  id?: string,
  callback?: (g: G) => void,
): G => {
  const ctx = new Context();
  const g = new cls(ctx, id);
  if (typeof callback === 'function') {
    callback(g);
  }
  if (strictMode) {
    g.strict = strictMode;
  }
  return g;
};

/** API for creating directional graph objects. */
export const digraph = builder(Digraph);

/** API for creating omnidirectional graph objects. */
export const graph = builder(Graph);

/** Provides a strict mode API. */
export const strict = {
  /** API for creating directional graph objects in strict mode. */
  digraph: builder(Digraph, true),
  /** API for creating omnidirectional graph objects in strict mode. */
  graph: builder(Graph, true),
};
