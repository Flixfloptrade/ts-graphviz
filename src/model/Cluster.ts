/* eslint-disable @typescript-eslint/no-use-before-define,max-classes-per-file */
import { attribute } from '../attribute';
import {
  ClusterType,
  EdgeTargetLike,
  ICluster,
  IClusterCommonAttributes,
  IEdge,
  INode,
  ISubgraph,
  Compass,
} from '../types';
import { AttributesBase } from './AttributesBase';
import { isEdgeTarget, ForwardRefNode, isEdgeTargetLike, Node } from './Node';
import { Edge } from './Edge';
import { IEdgeTarget } from '../types';
import { Attributes } from './Attributes';

/**
 * Base class for clusters.
 * @hidden
 */
export abstract class Cluster<T extends string> extends AttributesBase<T> implements ICluster<T> {
  /** Cluster ID */
  public id?: string;

  /** Comments to include when outputting with toDot. */
  public comment?: string;
  /** Indicates the type of cluster. */
  public abstract readonly type: ClusterType;
  /** Common attributes of objects in the cluster. */
  public abstract readonly attributes: Readonly<IClusterCommonAttributes>;

  /**
   * Nodes in the cluster.
   * @hidden
   */
  public nodes: Map<string, INode> = new Map();

  /**
   * Edges in the cluster.
   * @hidden
   */
  public edges: Set<IEdge> = new Set();

  /**
   * Subgraphs in the cluster.
   * @hidden
   */
  public subgraphs: Set<ISubgraph> = new Set();

  // private readonly internal = {
  //   nodes: ReadonlyMap
  // };

  /**
   * Add a Node to the cluster.
   */
  public addNode(node: INode): void {
    this.nodes.set(node.id, node);
  }

  /**
   * Add Edge to the cluster.
   */
  public addEdge(edge: IEdge): void {
    this.edges.add(edge);
  }

  /**
   * Add a Subgraph to the cluster.
   */
  public addSubgraph(subgraph: ISubgraph): void {
    this.subgraphs.add(subgraph);
  }

  /**
   * Check if the Node exists in the cluster.
   */
  public existNode(nodeId: string): boolean {
    return this.nodes.has(nodeId);
  }

  /**
   * Check if the Edge exists in the cluster.
   */
  public existEdge(edge: IEdge): boolean {
    return this.edges.has(edge);
  }

  /**
   * Check if the Subgraph exists in the cluster.
   */
  public existSubgraph(subgraph: ISubgraph): boolean {
    return this.subgraphs.has(subgraph);
  }

  /**
   * Create a Subgraph and add it to the cluster.
   */
  public createSubgraph(id?: string): ISubgraph {
    const graph = new Subgraph(id);
    this.subgraphs.add(graph);
    return graph;
  }

  /**
   * Remove Node from the cluster.
   */
  public removeNode(node: INode | string): void {
    this.nodes.delete(typeof node === 'string' ? node : node.id);
  }

  /**
   * Remove Edge from the cluster.
   */
  public removeEdge(edge: IEdge): void {
    this.edges.delete(edge);
  }

  /**
   * Remove Subgraph from the cluster.
   */
  public removeSubgraph(subgraph: ISubgraph): void {
    this.subgraphs.delete(subgraph);
  }

  /**
   * Create a Node in the cluster.
   */
  public createNode(id: string): INode {
    const node = new Node(id);
    this.nodes.set(id, node);
    return node;
  }

  /**
   * Get Subgraph in cluster by specifying id.
   *
   * If there is no Subgraph with the specified id in the cluster, return undefined.
   */
  public getSubgraph(id: string): ISubgraph | undefined {
    return Array.from(this.subgraphs.values()).find((subgraph) => subgraph.id === id);
  }

  /**
   * Get Node in cluster by specifying id.
   *
   * @description
   * If there is no Node with the specified id in the cluster, return undefined.
   */
  public getNode(id: string): INode | undefined {
    return this.nodes.get(id);
  }

  /** Create Edge and add it to the cluster. */
  public createEdge(targets: EdgeTargetLike[]): IEdge {
    if (targets.length < 2 && (isEdgeTargetLike(targets[0]) && isEdgeTargetLike(targets[1])) === false) {
      throw new Error('The element of Edge target is missing or not satisfied as Edge target.');
    }
    const edge = new Edge(targets.map((t) => this.toNodeLikeObject(t)));
    this.edges.add(edge);
    return edge;
  }

  /** @hidden */
  private toNodeLikeObject(node: EdgeTargetLike): IEdgeTarget {
    if (isEdgeTarget(node)) {
      return node;
    }
    const [id, port, compass] = node.split(':');
    const n = this.getNode(id);
    if (n !== undefined) {
      if (port && (compass === undefined || Compass.is(compass))) {
        return n.port({ port, compass });
      }
      return n;
    }
    if (Compass.is(compass)) {
      return new ForwardRefNode(id, { port, compass });
    }
    return new ForwardRefNode(id, { port });
  }

  /**
   * Declarative API for Subgraph.
   *
   * @description
   * If there is a Subgraph with the given ID, use it.
   * If not, create a Subgraph.
   *
   * @param id Subgraph ID.
   * @param callback Callback to operate Subgraph.
   */
  public subgraph(id?: string, callback?: (subgraph: ISubgraph) => void): ISubgraph {
    const subgraph: ISubgraph = id ? this.getSubgraph(id) ?? this.createSubgraph(id) : this.createSubgraph();
    if (callback) {
      callback(subgraph);
    }
    return subgraph;
  }

  /**
   * Declarative API for Node.
   *
   * @description
   * If there is a Node with the given ID, use it.
   * If not, create a Node.
   *
   * @param id Node ID.
   * @param callback Callback to operate Node.
   */
  public node(id: string, callback?: (node: INode) => void): INode {
    const node = this.getNode(id) ?? this.createNode(id);
    if (callback) {
      callback(node);
    }
    return node;
  }

  /**
   * Declarative API for Edge.
   *
   * @param targets Edges.
   * @param callback Callback to operate Edge.
   */
  public edge(targets: EdgeTargetLike[], callback?: (edge: IEdge) => void): IEdge {
    const edge = this.createEdge(targets);
    if (callback) {
      callback(edge);
    }
    return edge;
  }
}

/**
 * Subgraph object.
 * @category Primary
 */
export class Subgraph extends Cluster<attribute.Subgraph | attribute.ClusterSubgraph> implements ISubgraph {
  /** Indicates the type of cluster. */
  public type = ClusterType.subgraph;
  public attributes = {
    graph: new Attributes<attribute.ClusterSubgraph>(),
    edge: new Attributes<attribute.Edge>(),
    node: new Attributes<attribute.Node>(),
  };
  constructor(public readonly id?: string) {
    super();
  }
  /** Determines whether the Subgraph is a SubgraphCluster. */
  public isSubgraphCluster(): boolean {
    if (typeof this.id === 'string') {
      return this.id.startsWith('cluster');
    }
    return false;
  }
}
