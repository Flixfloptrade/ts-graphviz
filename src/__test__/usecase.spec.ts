import 'jest-graphviz';
import { Digraph, Graph } from '../model/cluster';
import { digraph, graph } from '../usecase';

describe('function digraph', () => {
  it('should return Digraph object, when execute digraph()', () => {
    const g = digraph();
    expect(g).toBeInstanceOf(Digraph);
  });

  test('callback style', () => {
    const G = digraph('G', g => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], e => {
        e.attributes.set('color', 'red');
      });
      g.subgraph('A', A => {
        const Aa = A.node('Aaa', n => {
          n.attributes.set('color', 'pink');
        });
        const Ab = A.node('Abb', n => {
          n.attributes.set('color', 'violet');
        });
        const Ac = A.node('Acc');
        A.edge([Aa.port('a'), Ab, Ac, 'E'], e => {
          e.attributes.set('color', 'red');
        });
      });
    });
    const dot = G.toDot();
    expect(dot).toMatchSnapshot();
    expect(dot).toBeValidDot();
  });
});

describe('function graph', () => {
  it('should return Graph object, when execute graph()', () => {
    const g = graph();
    expect(g).toBeInstanceOf(Graph);
  });

  test('callback style', () => {
    const G = graph('G', g => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], e => {
        e.attributes.set('color', 'red');
      });
      g.subgraph('A', A => {
        const Aa = A.node('Aaa', n => {
          n.attributes.set('color', 'pink');
        });
        const Ab = A.node('Abb', n => {
          n.attributes.set('color', 'violet');
        });
        const Ac = A.node('Acc');
        A.edge([Aa, Ab, Ac, A.node('hoge').port('fuga')], e => {
          e.attributes.set('color', 'red');
        });
      });
    });
    const dot = G.toDot();
    expect(dot).toMatchSnapshot();
    expect(dot).toBeValidDot();
  });
});
