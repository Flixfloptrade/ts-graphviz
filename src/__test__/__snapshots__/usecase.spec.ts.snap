// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function digraph callback style 1`] = `
"digraph \\"G\\" {
  \\"aa\\";
  \\"bb\\";
  \\"cc\\";
  subgraph \\"A\\" {
    \\"Aaa\\" [
      color = \\"pink\\",
    ];
    \\"Abb\\" [
      color = \\"violet\\",
    ];
    \\"Acc\\";
    \\"E\\";
    \\"Aaa\\":\\"a\\" -> \\"Abb\\" -> \\"Acc\\" -> \\"E\\" [
      color = \\"red\\",
    ];
  };
  \\"aa\\" -> \\"bb\\" -> \\"cc\\" [
    color = \\"red\\",
  ];
}"
`;

exports[`function graph callback style 1`] = `
"graph \\"G\\" {
  \\"aa\\";
  \\"bb\\";
  \\"cc\\";
  subgraph \\"A\\" {
    \\"Aaa\\" [
      color = \\"pink\\",
    ];
    \\"Abb\\" [
      color = \\"violet\\",
    ];
    \\"Acc\\";
    \\"hoge\\";
    \\"Aaa\\" -- \\"Abb\\" -- \\"Acc\\" -- \\"hoge\\":\\"fuga\\" [
      color = \\"red\\",
    ];
  };
  \\"aa\\" -- \\"bb\\" -- \\"cc\\" [
    color = \\"red\\",
  ];
}"
`;

exports[`function graph escape characters 1`] = `
"graph \\"G\\\\n\\\\\\"\\" {
  \\"a\\\\na\\";
  \\"b\\\\\\"b\\";
  \\"c\\\\nc\\\\\\"\\";
  \\"a\\\\na\\" -- \\"b\\\\\\"b\\" -- \\"c\\\\nc\\\\\\"\\" [
    color = \\"red\\",
  ];
}"
`;
